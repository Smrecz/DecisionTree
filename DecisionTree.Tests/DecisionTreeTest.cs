using System;
using System.Collections.Generic;
using System.Text.Json;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using DecisionTree.DotTreeExtensions;
using DecisionTree.Tests.Dto;
using DecisionTree.Tests.Model;
using DecisionTree.Tests.Tree;
using Xunit;

namespace DecisionTree.Tests
{
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("./approvals")]
    public class DecisionTreeTest
    {
        [Fact]
        public void DecisionTree_Should_Define_Graph()
        {
            //Arrange
            var tree = new ProjectDecisionTree();

            //Act
            var graphDefinition = tree.ConvertToDotGraph();

            //Assert
            Approvals.VerifyHtml(graphDefinition);
        }

        [Theory]
        [MemberData(nameof(DecisionTreeDtoList))]
        public void DecisionTree_Should_Follow_Paths(int counter, ItProjectDecisionDto dto)
        {
            //Arrange
            var tree = new ProjectDecisionTree();

            //Act
            tree.GetTrunk().Evaluate(dto);

            //Assert
            NamerFactory.AdditionalInformation = counter.ToString();
            Approvals.VerifyJson(JsonSerializer.Serialize(dto));
        }

        public static IEnumerable<object[]> DecisionTreeDtoList =>
           new List<object[]>
           {
               CreateDtoTestData(0,
                   new ItProject
                   {
                       ItemsToDo = 0
                   }),
               CreateDtoTestData(1,
                   new ItProject
                   {
                       ItemsToDo = 5,
                       IsOnHold = true
                   }),
               CreateDtoTestData(2,
                   new ItProject
                   {
                       ItemsToDo = 5,
                       Type = ProjectType.Internal
                   }),
               CreateDtoTestData(3,
                   new ItProject
                   {
                       ItemsToDo = 15,
                       Type = ProjectType.Financial,
                       TimeToDeadline = TimeSpan.FromDays(1)
                   }),
               CreateDtoTestData(4,
                   new ItProject
                   {
                       ItemsToDo = 5,
                       Type = ProjectType.Financial,
                       BudgetRemaining = 1000
                   }),
               CreateDtoTestData(5,
                   new ItProject
                   {
                       ItemsToDo = 15,
                       Type = ProjectType.Financial,
                       TimeToDeadline = TimeSpan.FromDays(10),
                       BudgetRemaining = 1000
                   }),
               CreateDtoTestData(6,
                   new ItProject
                   {
                       ItemsToDo = 15,
                       Type = ProjectType.Financial,
                       TimeToDeadline = TimeSpan.FromDays(10),
                       BudgetRemaining = 1000000
                   })
           };

        private static object[] CreateDtoTestData(int counter, ItProject project) =>
            new object[] {counter, new ItProjectDecisionDto { Project = project } };
    }
}
