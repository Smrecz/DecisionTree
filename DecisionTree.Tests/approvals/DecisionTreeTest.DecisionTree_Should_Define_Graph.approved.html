digraph G {
"dto => (dto.Project.ItemsToDo == 0)" -> "FinishResult" [label = "True"]
"FinishResult" [style = "filled" penwidth = 1 fillcolor = "#28a745" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#28a745"><tr><td bgcolor="#28a745" align="center"><font color="white">FinishResult</font></td></tr><tr><td bgcolor="white" cellpadding="1"></td></tr><tr><td align="left"><font color="white">dto =&gt; dto</font></td></tr><tr><td align="left"><font color="white">.SetResult(&quot;Project is finished.&quot;)</font></td></tr></table>>]
"dto => (dto.Project.ItemsToDo == 0)" -> "dto => dto.Project.IsOnHold" [label = "False"]
"dto => dto.Project.IsOnHold" -> "DoNothingResult" [label = "True"]
"DoNothingResult" [style = "filled" penwidth = 1 fillcolor = "#6c757d" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#6c757d"><tr><td bgcolor="#6c757d" align="center"><font color="white">DoNothingResult</font></td></tr></table>>]
"dto => dto.Project.IsOnHold" -> "dto => dto.Project.Type" [label = "False"]
"dto => dto.Project.Type" -> "DoNothingResult" [label = "Internal"]
"dto => dto.Project.Type" -> "dto => (dto.Project.ItemsToDo > 10)" [label = "#default_option"]
"dto => (dto.Project.ItemsToDo > 10)" -> "dto => (dto.Project.TimeToDeadline.Days < 7)" [label = "True"]
"dto => (dto.Project.TimeToDeadline.Days < 7)" -> "MoveDeadlineResult" [label = "True"]
"MoveDeadlineResult" [style = "filled" penwidth = 1 fillcolor = "#28a745" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#28a745"><tr><td bgcolor="#28a745" align="center"><font color="white">MoveDeadlineResult</font></td></tr><tr><td bgcolor="white" cellpadding="1"></td></tr><tr><td align="left"><font color="white">dto =&gt; dto</font></td></tr><tr><td align="left"><font color="white">.SetResult(&quot;Timeline reevaluation needed.&quot;)</font></td></tr></table>>]
"dto => (dto.Project.TimeToDeadline.Days < 7)" -> "dto => (dto.Project.BudgetRemaining < (dto.Project.ItemsToDo * 1000))" [label = "False"]
"dto => (dto.Project.BudgetRemaining < (dto.Project.ItemsToDo * 1000))" -> "RequestBudgetResult" [label = "True"]
"RequestBudgetResult" [style = "filled" penwidth = 1 fillcolor = "#28a745" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#28a745"><tr><td bgcolor="#28a745" align="center"><font color="white">RequestBudgetResult</font></td></tr><tr><td bgcolor="white" cellpadding="1"></td></tr><tr><td align="left"><font color="white">dto =&gt; dto</font></td></tr><tr><td align="left"><font color="white">.SetResult(&quot;Not enough funds.&quot;)</font></td></tr><tr><td align="left"><font color="white">.SetIsOnHold(True)</font></td></tr></table>>]
"dto => (dto.Project.BudgetRemaining < (dto.Project.ItemsToDo * 1000))" -> "DoNothingResult" [label = "False"]
"dto => (dto.Project.BudgetRemaining < (dto.Project.ItemsToDo * 1000))" [style = "filled" penwidth = 1 fillcolor = "#007bff" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#007bff"><tr><td bgcolor="#007bff" align="center"><font color="white">dto =&gt; (dto.Project.BudgetRemaining &lt; (dto.Project.ItemsToDo * 1000))</font></td></tr><tr><td bgcolor="white" cellpadding="1"></td></tr><tr><td align="left"><font color="white">dto =&gt; dto</font></td></tr><tr><td align="left"><font color="white">.SetIsBudgetReviewed(True)</font></td></tr></table>>]
"dto => (dto.Project.TimeToDeadline.Days < 7)" [style = "filled" penwidth = 1 fillcolor = "#17a2b8" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#17a2b8"><tr><td bgcolor="#17a2b8" align="center"><font color="white">dto =&gt; (dto.Project.TimeToDeadline.Days &lt; 7)</font></td></tr></table>>]
"dto => (dto.Project.ItemsToDo > 10)" -> "dto => (dto.Project.BudgetRemaining < (dto.Project.ItemsToDo * 1000))" [label = "False"]
"dto => (dto.Project.ItemsToDo > 10)" [style = "filled" penwidth = 1 fillcolor = "#17a2b8" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#17a2b8"><tr><td bgcolor="#17a2b8" align="center"><font color="white">dto =&gt; (dto.Project.ItemsToDo &gt; 10)</font></td></tr></table>>]
"dto => dto.Project.Type" [style = "filled" penwidth = 1 fillcolor = "#17a2b8" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#17a2b8"><tr><td bgcolor="#17a2b8" align="center"><font color="white">dto =&gt; dto.Project.Type</font></td></tr></table>>]
"dto => dto.Project.IsOnHold" [style = "filled" penwidth = 1 fillcolor = "#17a2b8" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#17a2b8"><tr><td bgcolor="#17a2b8" align="center"><font color="white">dto =&gt; dto.Project.IsOnHold</font></td></tr></table>>]
"dto => (dto.Project.ItemsToDo == 0)" [style = "filled" penwidth = 1 fillcolor = "#17a2b8" fontname = "Courier New" shape = "Mrecord" label = <<table border="0" cellborder="0" cellpadding="3" bgcolor="#17a2b8"><tr><td bgcolor="#17a2b8" align="center"><font color="white">dto =&gt; (dto.Project.ItemsToDo == 0)</font></td></tr></table>>]
}